1612/eb/tech-1191 ability to suppress honeybadger exceptions

zmathe-sv opened the pull request:
@mikeweaver can you please review? ([TECH-1191](https://ringrevenue.atlassian.net/browse/TECH-1191))
I moved the exception data logic to a single place so that it's shared between emails and honeybadger. The honebadger context now should support all the sections that were included in the emails.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/invoca/exception_handling/16)
<!-- Reviewable:end -->

mikeweaver commented:
LGTM other than two little issues.

@primiti Would you mind taking a second look?

---

Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[Gemfile.lock, line 16 at r1](https://reviewable.io:443/reviews/invoca/exception_handling/16#-KZODkYES22SphvB1uJQ:-KZODkYES22SphvB1uJR:berz1wz) ([raw file](https://github.com/invoca/exception_handling/blob/9393130a15f046353f101e3361bd043bc5e7444d/Gemfile.lock#L16)):*
> ```Ruby
>   specs:
>     actionmailer (3.2.19)
>       actionpack (= 3.2.19)
> ```

While we are here, we should take care of https://ringrevenue.atlassian.net/browse/TECH-1334: upgrade the honeybadger gem to the latest version. Dev effort should be 0. QA effort is the same as testing the changes that are already in this ticket.

---

*[lib/exception_handling.rb, line 116 at r1](https://reviewable.io:443/reviews/invoca/exception_handling/16#-KZOEtHBYzgKnli1_SZk:-KZOEtHBYzgKnli1_SZl:b-4m394u) ([raw file](https://github.com/invoca/exception_handling/blob/9393130a15f046353f101e3361bd043bc5e7444d/lib/exception_handling.rb#L116)):*
> ```Ruby
>     def log_error_rack(exception, env, rack_filter)
>       timestamp = set_log_error_timestamp
>       controller = env['action_controller.instance']
> ```

Could the exception info class extract the controller instance from the ENV if a controller is not passed explicitly?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/invoca/exception_handling/16#-:-KZOHB6QLRxBz7Ww6WbR:b-yb1gc1)*
<!-- Sent from Reviewable.io -->

zmathe-sv commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[Gemfile.lock, line 16 at r1](https://reviewable.io:443/reviews/invoca/exception_handling/16#-KZODkYES22SphvB1uJQ:-KZRn_KEXSONKLWCWhhz:b-ogzxew) ([raw file](https://github.com/invoca/exception_handling/blob/9393130a15f046353f101e3361bd043bc5e7444d/Gemfile.lock#L16)):*
<details><summary><i>Previously, mikeweaver (Mike Weaver) wrote…</i></summary><blockquote>

While we are here, we should take care of https://ringrevenue.atlassian.net/browse/TECH-1334: upgrade the honeybadger gem to the latest version. Dev effort should be 0. QA effort is the same as testing the changes that are already in this ticket.
</blockquote></details>

Ok, I'll update the version in Web and PNAPI repos. I'm not aware of any other places.
Will create the PRs after this CR is approved so that I'll have the proper SHA to use in Gemfiles. Btw, should I use the SHA coming from the merge into master, or it's ok to use the SHA from the last commit of this PR?

---

*[lib/exception_handling.rb, line 116 at r1](https://reviewable.io:443/reviews/invoca/exception_handling/16#-KZOEtHBYzgKnli1_SZk:-KZRnQSuuzYF9u5IVHHH:b-9jc936) ([raw file](https://github.com/invoca/exception_handling/blob/9393130a15f046353f101e3361bd043bc5e7444d/lib/exception_handling.rb#L116)):*
<details><summary><i>Previously, mikeweaver (Mike Weaver) wrote…</i></summary><blockquote>

Could the exception info class extract the controller instance from the ENV if a controller is not passed explicitly?
</blockquote></details>

Yes, I moved this logic to the info class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/invoca/exception_handling/16)*
<!-- Sent from Reviewable.io -->

ColinDKelley commented on lib/exception_handling.rb:
```diff
@@ -132,7 +131,7 @@ def log_error_rack(exception, env, rack_filter)
         # controller may not exist in some cases (like most 404 errors)
-        if controller
+        if (controller = exception_info.controller)
           controller.session["last_exception_timestamp"] = ExceptionHandling.last_exception_timestamp
```
I don't think `ExceptionHandling` is needed here. It would be more DRY to omit it.

ColinDKelley commented on test/unit/exception_handling/exception_info_test.rb:
```diff
@@ -134,6 +134,28 @@ class ExceptionInfoTest < ActiveSupport::TestCase
+        }
+
+        assert_equal_with_diff expected_data, prepare_data(exception_info.enhanced_data)
+      end
```
Is there a test for what `controller_from_context` does when the context is not a hash?

mikeweaver commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/invoca/exception_handling/16)*
<!-- Sent from Reviewable.io -->

primiti commented:
:+1: assuming you address Colin's comment.  Nice work!

zmathe-sv commented:
Review status: 10 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[lib/exception_handling.rb, line 135 at r2](https://reviewable.io:443/reviews/invoca/exception_handling/16#-KZODMDvRAvtFm_a-wR6-r2-135:-KZWhITfpFNUqFUJW_Ll:bxyy4uo) ([raw file](https://github.com/invoca/exception_handling/blob/8afeee531b1c0f5240a3cfc1d7c616f29062acda/lib/exception_handling.rb#L135)):*
<details><summary><i>Previously, ColinDKelley (Colin Kelley) wrote…</i></summary><blockquote>

I don't think `ExceptionHandling` is needed here. It would be more DRY to omit it.
</blockquote></details>

Sure, I have removed it.

---

*[test/unit/exception_handling/exception_info_test.rb, line 157 at r2](https://reviewable.io:443/reviews/invoca/exception_handling/16#-KZODMDvRAvtFm_a-wRG-r2-157:-KZWhmLWKjsUMRemxWD5:bofgwi9) ([raw file](https://github.com/invoca/exception_handling/blob/8afeee531b1c0f5240a3cfc1d7c616f29062acda/test/unit/exception_handling/exception_info_test.rb#L157)):*
<details><summary><i>Previously, ColinDKelley (Colin Kelley) wrote…</i></summary><blockquote>

Is there a test for what `controller_from_context` does when the context is not a hash?
</blockquote></details>

No, I've added new tests to cover that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/invoca/exception_handling/16)*
<!-- Sent from Reviewable.io -->

mikeweaver commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[Gemfile.lock, line 16 at r1](https://reviewable.io:443/reviews/invoca/exception_handling/16#-KZODkYES22SphvB1uJQ:-KZYAcDzRmsDHV4FmW4J:b-cumak7) ([raw file](https://github.com/invoca/exception_handling/blob/9393130a15f046353f101e3361bd043bc5e7444d/Gemfile.lock#L16)):*
<details><summary><i>Previously, zmathe-sv (Zsolt Mathe) wrote…</i></summary><blockquote>

Ok, I'll update the version in Web and PNAPI repos. I'm not aware of any other places.
Will create the PRs after this CR is approved so that I'll have the proper SHA to use in Gemfiles. Btw, should I use the SHA coming from the merge into master, or it's ok to use the SHA from the last commit of this PR?
</blockquote></details>

You need to use the SHA from merging into master. However, that should be a FF merge, so the SHA should not change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/invoca/exception_handling/16)*
<!-- Sent from Reviewable.io -->
